src/MathExpression.cpp

```diff
@@ -0,0 +1,181 @@
+#include "MathExpression.h"
+
+#include <cctype>
+#include <sstream>
+#include <set>
+#include <vector>
+
+#include "Result.h"
+
+
+
+MathExpression::MathExpression(std::string equation)
+: equation_(sanitizeEquation(equation)) {};
+
+
+
+std::set<std::string> MathExpression::getVariableNames() const
+{
+    std::set<std::string> variablesList;
+
+    std::string variableName = "";
+    for (int i = 0; i < equation_.length(); i++)
+    {
+        variableName = "";
+        if (isalpha(equation_[i]))
+        {
+            while (isalpha(equation_[i]) && i < equation_.length())
+            {
+                variableName += equation_[i++];
+            }
+            i--;
+            variablesList.insert(variableName);
+        }
+    }
+    return variablesList;
+};
+
+
+
+std::set<std::string> MathExpression::needsVariableValues() const
+{
+
+    std::set<std::string> variablesList = getVariableNames(); 
+    std::set<std::string> emptyVariablesList; 
+    for(auto it = variablesList.cbegin(); it != variablesList.end(); it++)
+    {
+        if (variables_.count(*it) == 0)
+        {
+            emptyVariablesList.insert(*it);
+        }
+    }
+
+    return emptyVariablesList;
+};
+
+
+
+void MathExpression::setVariableValue(std::string variableName, double value)
+{
+    variables_.erase  (variableName);
+    variables_.emplace(variableName, value);
+};
+
+
+
+std::string MathExpression::getRawEquation() const
+{
+    return equation_;
+};
+
+
+Result<std::vector<std::string>> MathExpression::getPopulatedEquation() const
+{
+    std::vector<std::string> populatedEquation;
+    for (int i = 0; i < equation_.length(); i++)
+    {
+        if (equation_[i] == '-')
+        {
+            if(i + 1 >= equation_.length() ||                                                                   // end of string
+               (i > 0 && (isalnum(equation_[i - 1]) || equation_[i - 1] == '.' || equation_[i - 1] == ')')) ||  // previous value supports a subtraction after it
+               (!isalnum(equation_[i + 1]) && equation_[i + 1] != '.'))                                         // next value doesn't support negation
+            {
+                populatedEquation.push_back(std::string(1, equation_[i]));
+            }
+            else if(isalpha(equation_[i + 1]))
```

Comment:
both this `else if(isalpha(equation_[i + 1]))` and `else if(isdigit(equation_[i + 1]) || equation_[i + 1] == '.')` blocks are almost the same as the blocks at 107 and 121. These should be pulled out into a single helper method for code reuse.

Reply:
I started pulling them apart (by incorporating the whole thing into extractNextNumberFromString/extractNextVariableFromString,) but ran into the issue of using the length of the returned strings to increment i; I could pass i in by reference, but that feels off to me.